src/main/scala/beam/router/skim/ODSkimmer.scala

```diff
@@ -0,0 +1,673 @@
+        correctedTrip.costEstimate,
+        1,
+        energyConsumption
+      )
+    currentSkim.get(key) match {
+      case Some(existingSkim: ODSkimmerInternal) =>
+        val count = existingSkim.numObservations
+        val newPayload = ((existingSkim * count + payload) / (count + 1)).asInstanceOf[ODSkimmerInternal]
+        currentSkim.put(key, newPayload.copy(numObservations = count + 1))
+      case _ =>
+        currentSkim.put(key, payload)
+    }
+  }
+
+  private def getExcerptData(
+    timePeriodString: String,
+    hoursIncluded: List[Int],
+    origin: TAZ,
+    destination: TAZ,
+    mode: BeamMode,
+    dummyId: Id[BeamVehicleType]
+  ): ExcerptData = {
+    val individualSkims = hoursIncluded.map { timeBin =>
+      currentSkim
+        .get(ODSkimmerKey(timeBin, mode, origin.tazId, destination.tazId))
+        .map(_.asInstanceOf[ODSkimmerInternal].toSkimExternal)
+        .getOrElse {
+          val adjustedDestCoord = if (origin.equals(destination)) {
+            new Coord(
+              origin.coord.getX,
+              origin.coord.getY + Math.sqrt(origin.areaInSquareMeters) / 2.0
+            )
+          } else {
+            destination.coord
+          }
+          getSkimDefaultValue(
+            mode,
+            origin.coord,
+            adjustedDestCoord,
+            timeBin * 3600,
+            dummyId,
+            beamServices
+          )
+        }
+    }
+    val weights = individualSkims.map(sk => Math.max(sk.count, 1).toDouble)
+    val sumWeights = weights.sum
+    val weightedDistance = individualSkims.map(_.distance).zip(weights).map(tup => tup._1 * tup._2).sum / sumWeights
+    val weightedTime = individualSkims.map(_.time).zip(weights).map(tup => tup._1 * tup._2).sum / sumWeights
+    val weightedGeneralizedTime = individualSkims
+      .map(_.generalizedTime)
+      .zip(weights)
+      .map(tup => tup._1 * tup._2)
+      .sum / sumWeights
+    val weightedCost = individualSkims.map(_.cost).zip(weights).map(tup => tup._1 * tup._2).sum / sumWeights
+    val weightedGeneralizedCost = individualSkims
+      .map(_.generalizedCost)
+      .zip(weights)
+      .map(tup => tup._1 * tup._2)
+      .sum / sumWeights
+    val weightedEnergy = individualSkims.map(_.energy).zip(weights).map(tup => tup._1 * tup._2).sum / sumWeights
+
+    ExcerptData(
+      timePeriodString = timePeriodString,
+      mode = mode,
+      originTazId = origin.tazId,
+      destinationTazId = destination.tazId,
+      weightedTime = weightedTime,
+      weightedGeneralizedTime = weightedGeneralizedTime,
+      weightedCost = weightedCost,
+      weightedGeneralizedCost = weightedGeneralizedCost,
+      weightedDistance = weightedDistance,
+      sumWeights = sumWeights,
+      weightedEnergy = weightedEnergy
+    )
+  }
+
+}
+
+object ODSkimmer extends LazyLogging {
+
+  private var pastSkims: immutable.List[immutable.Map[ODSkimmerKey, ODSkimmerInternal]] = immutable.List()
```

Comment:
These should be a `mutable.List`, as an immutable var is generally considered the worst option. Also - why are these being used in a static way? They probably should be atomic if needed, but I'd prefer to contain it in the class itself...although it could still be a concurrency problem.